feat(CalibrationStatusClassification): add calibration status classification

amadeuszsz opened the pull request:
## Summary

A new project for miscalibration detection:
* The model consumes 5D array (R, G, B, depth, intensity) obtained by fusion of image, point cloud and projection with use of provided transform between camera and LiDAR.
* The model outputs an array of two elements [SCORE_MISCALIBRATED, SCORE_CALIBRATED].

## Change point

* Added CalibrationStatusClassification project.
* Added new transforms for dataset.

## Note

* For now, dataset available in [TIER IV INTERNAL LINK](https://drive.google.com/drive/folders/1PiTj_LuFO57C-zcct0KOYjLQ-mkZssZ5).
* Train results in [TIER IV INTERNAL LINK](https://drive.google.com/file/d/1U8zLGpdnI7nHx17BjIoULA95I1p3mp60/view?usp=drive_link).
* ONNX in [TIER IV INTERNAL LINK](https://drive.google.com/file/d/103jQ8VbgjgcGwvc2fU2YJOYjO_uJdcDV/view?usp=drive_link).
* Instructions already included in README (training and deployment).
* TODO in next PRs:
  - Update config because current version results with training instability (during tests now).
  - Enhance preprocessing with respect to incoming inference c++ node. Algorithm has to match in training and inference.
  - Data generation pipeline.

## Test performed

- Log

```
(container) autoware@npc2405014:/workspace (feat/calibration-status-classification *%)$ python3 projects/CalibrationStatusClassification/deploy/main.py projects/CalibrationStatusClassification/configs/deploy/resnet18_5ch.py projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_t4base.py work_dirs/resnet18_5ch_1xb8-25e_t4base/epoch_25.pth data/calibrated_data/validation_set/data/0_image.jpg  --device cuda:0 --work-dir /workspace/work_dirs/ --verify
INFO:mmdeploy:Loading deploy config from: projects/CalibrationStatusClassification/configs/deploy/resnet18_5ch.py
INFO:mmdeploy:Loading model config from: projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_t4base.py
INFO:mmdeploy:Loading model from checkpoint: work_dirs/resnet18_5ch_1xb8-25e_t4base/epoch_25.pth
Loads checkpoint by local backend from path: work_dirs/resnet18_5ch_1xb8-25e_t4base/epoch_25.pth
INFO:mmdeploy:Loading sample data from: data/calibrated_data/validation_set/data/0_image.jpg
INFO:mmdeploy:Exporting model to ONNX format...
INFO:mmdeploy:Input shape: torch.Size([1, 5, 1860, 2880])
INFO:mmdeploy:Output path: /workspace/work_dirs/end2end.onnx
INFO:mmdeploy:ONNX opset version: 16
INFO:mmdeploy:ONNX export completed: /workspace/work_dirs/end2end.onnx
INFO:mmdeploy:Simplifying ONNX model...
INFO:mmdeploy:ONNX model simplified successfully. Saved to /workspace/work_dirs/end2end.onnx
INFO:mmdeploy:Converting ONNX to TensorRT...
INFO:mmdeploy:Setting input shapes - min: [1, 5, 1860, 2880], opt: [1, 5, 1860, 2880], max: [1, 5, 1860, 2880]
INFO:mmdeploy:Successfully parsed the ONNX file
INFO:mmdeploy:Building TensorRT engine...
INFO:mmdeploy:TensorRT engine saved to /workspace/work_dirs/end2end.engine
INFO:mmdeploy:Engine max workspace size: 1.00 GB
INFO:mmdeploy:TensorRT conversion successful: /workspace/work_dirs/end2end.engine
INFO:mmdeploy:Running verification for miscalibrated and calibrated samples with an output array [SCORE_MISCALIBRATED, SCORE_CALIBRATED]...
INFO:mmdeploy:==================================================
INFO:mmdeploy:Verifying miscalibrated sample...
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying PyTorch model...
INFO:mmdeploy:PyTorch inference latency: 1285.16 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  Output: [[ 2.993775  -2.9912994]]
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying ONNX model...
INFO:mmdeploy:ONNX inference latency: 826.14 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  ONNX output: [[ 2.993775  -2.9912994]]
INFO:mmdeploy:  Max difference with PyTorch: 0.000000
INFO:mmdeploy:  Mean difference with PyTorch: 0.000000
INFO:mmdeploy:ONNX model verification passed!
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying TensorRT model...
INFO:mmdeploy:TensorRT inference latency: 37.72 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  TensorRT output: [[ 2.9937606 -2.9912848]]
INFO:mmdeploy:  Max difference with PyTorch: 0.000015
INFO:mmdeploy:  Mean difference with PyTorch: 0.000014
INFO:mmdeploy:TensorRT model verification passed!
INFO:mmdeploy:==================================================
INFO:mmdeploy:Verifying calibrated sample...
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying PyTorch model...
INFO:mmdeploy:PyTorch inference latency: 1240.64 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  Output: [[-4.277913  4.276496]]
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying ONNX model...
INFO:mmdeploy:ONNX inference latency: 824.78 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  ONNX output: [[-4.277914  4.276497]]
INFO:mmdeploy:  Max difference with PyTorch: 0.000001
INFO:mmdeploy:  Mean difference with PyTorch: 0.000001
INFO:mmdeploy:ONNX model verification passed!
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying TensorRT model...
INFO:mmdeploy:TensorRT inference latency: 36.39 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  TensorRT output: [[-4.277993  4.276578]]
INFO:mmdeploy:  Max difference with PyTorch: 0.000082
INFO:mmdeploy:  Mean difference with PyTorch: 0.000081
INFO:mmdeploy:TensorRT model verification passed!
INFO:mmdeploy:==================================================
INFO:mmdeploy:Deployment completed successfully!
```

SamratThapa120 reviewed (approved):
Thanks for the great PR and initiative to add deep learning based camera-lidar calibration checks.

Overall the PR looks good to me. In the future, maybe we could increase the training dataset size by using these [T4 datasets ](https://github.com/tier4/AWML/tree/main/autoware_ml/configs/t4dataset) for training. Although some of the datasets can contain slight miscalibrations, most of them can be used .

KSeangTan commented on autoware_ml/classification2d/datasets/transforms/calibration_classification_transform.py:
```diff
@@ -0,0 +1,378 @@
+        # Visualize the results
+        if self.debug:
+            self.visualize_projection(input_data, label, camera_data, undistorted_data)
+            # self.visualize_results(input_data, label, camera_data, undistorted_data)
```
remove?

amadeuszsz replied:
Addressed in https://github.com/tier4/AWML/pull/67/commits/92ab7fb95d2c71d52bfc3828207d70ab44ff1dd6 for all the commented code.

KSeangTan commented on autoware_ml/classification2d/datasets/transforms/calibration_classification_transform.py:
```diff
@@ -0,0 +1,378 @@
+
+        return resized_image, calibration_data
+
+    def apply_affine_transformation(self, image, calibration_data, max_distortion=0.001):
```
`calibration_data` is unused?

amadeuszsz replied:
Addressed in https://github.com/tier4/AWML/pull/67/commits/43bee94a9294b49c5adf0fda0391467aa6e7eb26 for all the unused variables.

KSeangTan commented on autoware_ml/classification2d/datasets/transforms/calibration_classification_transform.py:
```diff
@@ -0,0 +1,378 @@
+            np.random.seed(None)
+
+        # Loading and preparing data
+        camera_data, lidar_data, calibration_data = self.load_data(results["img_path"])
```
`calibration_data` is used a lot and there's few `keys` in the dict. In this case, I would recommend to make a `NamedTuple` or `dataclass` for it. Otherwise, it's not clean about the content in the `dict`.

amadeuszsz replied:
Good idea! Addressed in https://github.com/tier4/AWML/pull/67/commits/46226c5b3988140e2d7a6a7cefba8305fd102e3e

KSeangTan commented on autoware_ml/classification2d/datasets/transforms/calibration_classification_transform.py:
```diff
@@ -0,0 +1,378 @@
+        if self.test and force_generate_miscalibration:
+            raise ValueError("force_generate_miscalibration is not supported in test mode")
+        if self.test:  # in test case, label is not used but must be registered
+            generate_miscalibration = False
```
```
generate_miscalibration = False if self.test else force_generate_miscalibration or random.choice([True, False]) # 50/50 split
```

amadeuszsz replied:
This is just a change to one liner 😄 In this file, for consistency purpose I prefer to keep as it is.

KSeangTan commented on autoware_ml/classification2d/datasets/transforms/calibration_classification_transform.py:
```diff
@@ -0,0 +1,378 @@
+
+
+@TRANSFORMS.register_module()
+class CalibrationClassificationTransform(BaseTransform):
```
Is it possible to modularize this class to different augmentation phases and make it a cleaner augmentation pipeline?

KSeangTan replied:
For example, 
`undistort_image` might be used by another project in the future

KSeangTan reviewed (changes requested):
LGTM overall
Would you mind addressing some comments? Thanks

amadeuszsz commented on autoware_ml/classification2d/datasets/transforms/camera_lidar_augmentation.py:
```diff
@@ -0,0 +1,107 @@
+    return int(255 * r), int(255 * g), int(255 * b)
+
+
+def bounded_gaussian(center, min_value, max_value, scale):
```
Thanks!
Addressed in https://github.com/tier4/AWML/pull/67/commits/6bf2c6dc57fdecca6cb5d8f8f6985124da7ee57c

amadeuszsz commented on autoware_ml/classification2d/datasets/transforms/camera_lidar_augmentation.py:
```diff
@@ -0,0 +1,107 @@
+import transforms3d
+
+
+def intensity_to_rainbow_color(value):
```
Addressed in https://github.com/tier4/AWML/pull/67/commits/8e5c9ce1f31ff4ab457e71535547c6f49dec41d4 for all the related code.

amadeuszsz commented:
@KSeangTan 
I changed my mind - in order to speed up development, we would like to process with merging this PR as it is, therefore I addressed all your comments. Remaining improvements we will address in separate PRs.

vividf reviewed (approved):
Thanks for your PR!
LGTM!
I will handle data generation and load the dataset using the AWML method.

KSeangTan commented on projects/CalibrationStatusClassification/deploy/main.py:
```diff
@@ -0,0 +1,562 @@
+) -> bool:
+    """Export ONNX model to TensorRT format."""
+    try:
+        import tensorrt as trt
```
Better if we can move `import` to the top, and check import error there

amadeuszsz replied:
Addressed in https://github.com/tier4/AWML/pull/67/commits/96aec9e9e25700624109288118e34f1fbd11421f

KSeangTan commented on tools/classification2d/dist_train.sh:
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
```
Any reason we cannot use the script from `detection3d`?

KSeangTan commented on autoware_ml/classification2d/datasets/transforms/calibration_classification_transform.py:
```diff
@@ -0,0 +1,520 @@
+        if generate_miscalibration:
+            calibration_data.camera_to_lidar_pose = alter_calibration(
+                calibration_data.camera_to_lidar_pose,
+                min_augmentation_angle=1.0,
```
Is it better if we make these values parameters?

amadeuszsz replied:
Next PR from my side will probably rework preprocessing, so for now please let me leave it as it is 🙏🏻

KSeangTan commented on autoware_ml/classification2d/datasets/transforms/calibration_classification_transform.py:
```diff
@@ -0,0 +1,520 @@
+        max_intensity = intensities.max()
+        if min_intensity == max_intensity:
+            return np.zeros_like(intensities)
+        return (intensities - min_intensity) / (max_intensity - min_intensity)
```
Maybe mention the normalization is done with min-max normalization in docstring

amadeuszsz replied:
Addressed in https://github.com/tier4/AWML/pull/67/commits/cf23e8ae389255cc6995d333c5c652ee7779c9ab

amadeuszsz commented on autoware_ml/classification2d/datasets/transforms/camera_lidar_augmentation.py:
```diff
@@ -0,0 +1,147 @@
+    return transforms3d.quaternions.quat2mat((w, factor * x, factor * y, factor * z))
+
+
+def random_translation(radius: float) -> np.ndarray:
```
Addressed in https://github.com/tier4/AWML/pull/67/commits/054d919e0fea28fad8a18c4e3f16ddced4a9ceb3

amadeuszsz commented on projects/CalibrationStatusClassification/deploy/main.py:
```diff
@@ -0,0 +1,562 @@
+            logger.warning("ONNX model simplification failed")
+    except ImportError:
+        logger.info("onnxsim not installed, skipping model simplification")
+    except Exception as e:
```
Addressed as a part of https://github.com/tier4/AWML/pull/67/commits/6be45d38b79344213eab21915b7ce5ab0243807c

amadeuszsz commented on projects/CalibrationStatusClassification/deploy/main.py:
```diff
@@ -0,0 +1,562 @@
+    logger.info("Deployment completed successfully!")
+
+
+if __name__ == "__main__":
```
Addressed in https://github.com/tier4/AWML/pull/67/commits/43792ede650169ebb180b658b8e017fae5970cfb

Looking into script we can definitely make it reusable for AWML 👍🏻
